services/121-service/src/transaction-job-processors/transaction-job-processors.service.ts

```diff
@@ -224,18 +228,119 @@ export class TransactionJobProcessorsService {
+    }
+
+    // 2. Check if all required properties are present. If not, create a failed transaction and throw an error.
+    for (const [name, value] of Object.entries(input)) {
+      // TODO: make some properties optional like in Visa, but why?
+
+      // Define "empty" based on your needs. Here, we check for null, undefined, or an empty string.
+      if (value === null || value === undefined || value === '') {
+        const errorText = `Property ${name} is undefined`;
+        await this.createTransactionAndUpdateRegistration({
+          programId: input.programId,
+          paymentNumber: input.paymentNumber,
+          userId: input.userId,
+          calculatedTransferAmountInMajorUnit: input.transactionAmount,
+          financialServiceProviderId: financialServiceProvider.id,
+          registration,
+          oldRegistration,
+          isRetry: input.isRetry,
+          status: StatusEnum.error,
+          errorText,
+        });
+        return;
+      }
+    }
+
+    // 3. Start the transfer, save error transaction on failure
+    let safaricomDoTransferResult: DoTransferReturnParams;
+    try {
+      safaricomDoTransferResult = await this.safaricomService.doTransfer({
+        transactionAmount: input.transactionAmount,
+        programId: input.programId,
+        paymentNr: input.paymentNumber,
+        userId: input.userId,
+        referenceId: input.referenceId,
+        registrationProgramId: input.registrationProgramId,
+        phoneNumber: input.phoneNumber,
+        nationalId: input.nationalId,
+      });
+    } catch (error) {
+      await this.createTransactionAndUpdateRegistration({
+        programId: input.programId,
+        paymentNumber: input.paymentNumber,
+        userId: input.userId,
+        calculatedTransferAmountInMajorUnit: input.transactionAmount,
+        financialServiceProviderId: financialServiceProvider.id,
+        registration,
+        oldRegistration,
+        isRetry: input.isRetry,
+        status: StatusEnum.error,
+        errorText: error?.message,
+      });
+      return;
+    }
+
+    // 4. If transfer is successful, create message and add to queue (not needed for safaricom)
+
+    // 5. create success transaction and update registration
+    const transaction = await this.createTransactionAndUpdateRegistration({
+      programId: input.programId,
+      paymentNumber: input.paymentNumber,
+      userId: input.userId,
+      calculatedTransferAmountInMajorUnit:
+        safaricomDoTransferResult.amountTransferredInMajorUnit,
+      financialServiceProviderId: financialServiceProvider.id,
+      registration,
+      oldRegistration,
+      isRetry: input.isRetry,
+      status: StatusEnum.waiting, // This will only go to 'success' via callback
+    });
+
+    // 6. Storing safaricom transfer data (new compared to visa)
+    // TODO: refactor/move this up, so that this is also saved on error transactions.
+    await this.safaricomService.createAndSaveSafaricomTransferData(
```

Comment:
Refactor into using the (custom) repo directly, see sequence diagram Miro.

The Safaricom module cannot know about transactions, so cannot use a transaction as a parameter.

Futhermore, function signatures should only contain the values needed and be explicit. See guidelines.